### ISSUE TITLE
Returns wrong number of posts when we have password-protected posts.

### ISSUE THREAD

**Issue by miya0001** [WP-API/WP-API#2299]
If we have password-protected posts, API returns posts without password-protected posts.

In this case, API will returns 9 posts when we have a password-protected post, 

```
 {
    per_page: 10,
    offset: 0
}
```

Then we send request like fowllowing.
The first post in response is duplicate.

```
 {
    per_page: 10,
    offset: 9
}
```

I know I always should set `offset : 10`, but I feel It will make us confuse.

**Comment by kadamwhite** [comment]
I think this (used to?) happen with sticky posts as well, which are excluded by default -- I do agree that it's unintuitive, if I ask for 10 posts I would expect to get 10 back. What's the rationale for not treating pass-protected posts the same way we'd treat any other non-matching filter, I.e. Return the collection as if it did not exist?

**Comment by danielbachhuber** [comment]
This is a known issue I've been meaning to create a ticket about.

The problem is that, across all of our collections, we perform a final permissions check before we add the item to the result set. In this case, the password-protected post fails the permissions check, so it's removed from the result set. See #1055 for ongoing problems with password-protected posts.

However, the problem isn't specific to password-protected posts — because capability checks can be conditionally filtered, _any_ post could be removed from the result set at a given time. Because it's a runtime filter, there's no way to accommodate for it in our original SQL query.

Furthermore, we can't simply query for an additional post and add it to the set for a couple of reasons:
1. Because we'd be departing from the original SQL query, paginating the response would require querying for _all_ data and figuring out which page we were on in PHP.
2. Again, because filters happen at runtime, the capability filters could (conceivably) change between requests.

We do need to "solve" this problem in some way though — problem being that the client expects to receive the number of items they requested. Maybe we need to return a null object to represent posts the request doesn't have access to.

**Comment by miya0001** [comment]
> Maybe we need to return a null object to represent posts the request doesn't have access to.

:+1:

**Comment by adamsilverstein** [comment]
We should match the core behaviour here, which I am not certain of without testing and some digging. My expectation would be that if I requested 10 posts, I would get 10 posts. The unavailable to me posts (not sticky when i requested sticky, private/password protected posts, etc.) should ideally be excluded from the query entirely.

**Comment by rmccue** [comment]
This is obsoleted by #2701.

**Comment by danielbachhuber** [comment]
@rmccue What about my comment? https://github.com/WP-API/WP-API/issues/2299#issuecomment-187185707

**Comment by rmccue** [comment]
> Again, because filters happen at runtime, the capability filters could (conceivably) change between requests.

I discussed this with @joehoyle; I think if you're massively changing the permissions, you may be voiding your warranty. Missed that it was in this ticket though, sorry about that.

I'm not certain about using a `null`; it requires double-checking on the client end, as the type for array items is now a union type of `Post|null`. I think it'd better to leave the items out in this case.

Reopening to finish discussing that.

**Comment by kadamwhite** [comment]
I'm 👎  on `null` because it complicates iteration... but it's a mess of a situation that omitting it entirely would void the pagination counts.

> The unavailable to me posts (not sticky when i requested sticky, private/password protected posts, etc.) should ideally be excluded from the query entirely.

I understand why we can't have this nice thing, but @adamsilverstein articulates well the assumptions I would have made.

**Comment by rmccue** [comment]
I think basically the only case where this occurs is when plugins are overriding a meta cap, in which case, you're going to end up with all sorts of weird behaviour anyway. I think it should be up to those plugins to alter the query to exclude the posts.

An analogous behaviour would be if you change these meta caps, what happens to your normal posts page that's meant to show 10 posts? It'd be your responsibility to hook in to `pre_get_posts` to remove the posts.

**Comment by kadamwhite** [comment]
Resolved by #2720
